api-proposal.md

```diff
@@ -1,90 +1,152 @@
-# Proposal for ecosystem website api.
+# Proposal for Ecosystem Website API
 
 ## API's
 
 ---
 
-### Package List
+### Error handling
 
-We could have some specifc string matching for categories and tags in the query if thats what the user is searching for to maybe speed things up. `cat:<category>` and `tag:<string>`.
+On error, each endpoint returns a HTTP status code 500, and the following body
+```javascript
+data: {
+  error: string // containing an error message to be displayed in the UI
+}
+```
 
-- GET `/api/v1/packages?query=:string&page=:number`
-  ```
-  items: [<package>]
-  ```
+### Authentication
 
----
+An authenticated user is authenticated against GitHub.
+An authenticated administrator is defined by a set of GitHub usernames in the configuration of the backend.
+
+### Endpoints
 
-### Package
+#### Package
 
 - **Schema**
 
-  ```
+  ```javascript
   <package> {
-    id: number
-    name: string
-    description: string
-    readme: string // maybe from github or something?
-    image: blob || string
+    id: string // a unique, immutable string identifiying this package. allowed: [a-z-_]{2,}
+    name: string // human readable name
+    description: string 
+    readme: string // user can enter a markdown description
+    image: blob // we should restrict the image size to 5kb
     website: string
     repository: string
     license: string
     commentsCount: number
     upvotes: number
     downvotes: number
-    tags: string[]
+    tags: string[] // tags are created automatically if they don't exist yet. Only [a-z-_]{2,} is allowed for tags.
     added: datetime
     updated: datetime
+    owner: string // github username
   }
 
   ```
 
-- GET `/api/v1/packages/:package`
-  ```
+- **Get a package by id**
+
+  GET `/api/v1/packages/:package`
+  ```javascript
   data: {
     package: <package>
-    comments: [<comment>] // first 50(?) comments
-    releases: [<release>] // first 10(?) releases
```

Comment:
Is it the job of the frontend to make 3 ajax request for loading a package then?

Reply:
I somehow feel this is a better separation of concerns. Otherwise, paginating through comments or releases will be cumbersome or require separate logic.
I don't think the page's performance will suffer from the three calls.

Reply:
It won't, I'm just making sure I understand what the job of the front end is.

Reply:
Okay, I see :)